tests/test_writer.py

```diff
@@ -777,7 +777,7 @@ def test_colors_in_outline_item():
     reader2 = PdfReader(target)
     for outline_item in reader2.outline:
         # convert float to string because of mutability
-        assert [str(c) for c in outline_item.color] == [str(p) for p in purple_rgb]
+        assert [f"{c:.5f}" for c in outline_item.color] == [f"{p:.5f}" for p in purple_rgb]
```

Comment:
Why was this change necessary? What is getting outputted for `str(c)` and `str(p)` vs this new formatting?

Reply:
Good question! "Hidden" in this test was the assumption that floats are represented with 5 decimal places, because they're baked into the `purple_rgb` assignment:

```
    purple_rgb = (0.50196, 0, 0.50196)
```

However, since a `FloatObject` gets turned into a decimal, the in-memory representation will become:

```
>>> Decimal(0.50196)
Decimal('0.501959999999999961772800816106610000133514404296875')
```

So this patch uses the `.5f` specifier to explicitly match the `purple_rgb` assignment.

Reply:
Okay, so we could keep `str(p)` and then only apply the formatting to values coming from `c`. 

If we were to set `purple_rgb == ('0.50196', 0, '0.50196')`, then would we be able to keep the existing test asserts as-is?

Though, this is surprising to me, as it looks like we should always be casting the value to a string _before_ we convert to a decimal, so we should be doing:
```
>>> a = Decimal('0.50196')
>>> a
Decimal('0.50196')
>>> str(a)
'0.50196'
>>> '{0:f}'.format(a)
'0.50196'

Reply:
> Okay, so we could keep `str(p)` and then only apply the formatting to values coming from `c`.

Almost, but not quite, because the second value in that tuple (zero) does not have 5 decimal places:

```
E           AssertionError: assert ['0.50196', '0.00000', '0.50196'] == ['0.50196', '0', '0.50196']
E             At index 1 diff: '0.00000' != '0'
E             Use -v to get more diff
```

We'd be violating the type constraint if we used strings in the tuple:

```
color: Optional[Union[Tuple[float, float, float], str]] = None,
```

I think I see where you're going using strings, though. The test still fails with that change, so need to do a little more digging to find out why.

Reply:
Dug into this one, too.

The test passes the third case if we use strings in the `purple_rgb =('0.50196', 0, '0.50196')` assignment, however, it fails the first and second cases:

```python
    writer.add_outline_item("First Outline Item", pagenum=2, color="800080") # FAIL
    writer.add_outline_item("Second Outline Item", pagenum=3, color="#800080") # FAIL
    writer.add_outline_item("Third Outline Item", pagenum=4, color=purple_rgb) # PASS
```

The reason for this is when color is pass in as a hex string, it gets converted to a tuple of floats by `hex_to_rgb`:

```python
>>> from PyPDF2.generic import hex_to_rgb, FloatObject
>>> rgb = hex_to_rgb('#800080')
>>> rgb[0]
0.5019607843137255
>>> repr(FloatObject(rgb[0]))
'0.5019607843137254832299731788225471973419189453125'
```

So this PR does create a side effect of FloatObjects produced by `hex_to_rgb` having very high precision. One way we could deal with that is to have `hex_to_rgb` change its output from a tuple of floats to a tuple of Decimals so that we can set the precision to something reasonable like 5 decimal places.

Reply:
@MasterOdin That is a much better idea, and will be easy to include in unit tests. I'll take a stab at it.

Reply:
Just found out something super interesting about python's Decimal:

```python
>>> import decimal
>>> decimal.getcontext()
Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])
```

So you can call `decimal.getcontext().precision` to get the current precision. By default it's 28 (on my machine).

You can also do:

```python
with decimal.localcontext() as ctx:
  ctx.precision = 5
  # Do stuff with Decimal
```

By deferring to the decimal context, we won't need an extra parameter on `FloatObject`. Going to try it...

Reply:
Implemented your suggestion, which allowed us to keep the unit test for the writer the same, and will prevent the issue with floating converted from hex values having very high precision.

However, in the future, a better solution may be to do something similar to [decimal.Context](https://github.com/python/cpython/blob/main/Lib/_pydecimal.py#L435). We wouldn't need the extra argument on FloatObject anymore, and it'd give users more control over how formatting occurs.

Roughly, it could look something like:

```python
import PyPDF2
from PyPDF2 import PdfReader, PdfWriter
from PyPDF2.context import StripExtraTrailingZeros, QuantizeInteger

with PyPDF2.localcontext() as ctx:
  ctx.prec = 5 # specify precision
  ctx.flags = [StripExtraTrailingZeros, QuantizeInteger] # specify additional format flags

  reader = PdfReader("./path/to/file.pdf")
  writer = PdfWriter()

  # context settings will be used for formatting
```

Reply:
Modified the implementation a bit to make it simpler.

It seemed wrong to add a precision property to `FloatObject` since it inherits from `Decimal`, which already represents precision internally. So instead of adding an extra precision property to `FloatObject`, we can just pass it a "quantized" `Decimal` e.g. `d.quantize(Decimal('1.00000'))` sets the precision of `d` to 5.

If you agree, happy to rebase and squash the latest commit.